ee/presencedetection/presencedetection.go

```diff
@@ -0,0 +1,51 @@
+package presencedetection
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+const DetectionFailedDurationValue = -1 * time.Second
+
+type PresenceDetector struct {
+	lastDetectionUTC          time.Time
+	mutext                    sync.Mutex
+	hadOneSuccessfulDetection bool
+	// detectFunc that can be set for testing
+	detectFunc func(string) (bool, error)
+}
+
+// DetectPresence checks if the user is present by detecting the presence of a user.
+// It returns the duration since the last detection.
+func (pd *PresenceDetector) DetectPresence(reason string, detectionInterval time.Duration) (time.Duration, error) {
+	pd.mutext.Lock()
+	defer pd.mutext.Unlock()
+
+	if pd.detectFunc == nil {
+		pd.detectFunc = Detect
+	}
+
+	// Check if the last detection was within the detection interval
+	if pd.hadOneSuccessfulDetection && time.Since(pd.lastDetectionUTC) < detectionInterval {
+		return time.Since(pd.lastDetectionUTC), nil
+	}
+
+	success, err := pd.detectFunc(reason)
+
+	switch {
+	case err != nil && pd.hadOneSuccessfulDetection:
+		return time.Since(pd.lastDetectionUTC), fmt.Errorf("detecting presence: %w", err)
```

Comment:
This is a weird case. Do we want it?

Reply:
Yea I agree, but worth nothing that we are implicitly saying if a detection fails, we behave as if no detection in the past has ever been successful.